subsys/bluetooth/controller/ll_sw/ull_llcp_conn.c

```diff
@@ -0,0 +1,781 @@
+	tx_enqueue(conn, tx);
+}
+
+static void rp_cu_ntf(struct ull_cp_conn *conn, struct proc_ctx *ctx)
+{
+	struct node_rx_pdu *ntf;
+	struct node_rx_cu *pdu;
+
+	/* Allocate ntf node */
+	ntf = ntf_alloc();
+	LL_ASSERT(ntf);
+
+	ntf->hdr.type = NODE_RX_TYPE_CONN_UPDATE;
+	pdu = (struct node_rx_cu *)ntf->pdu;
+
+	pdu->status = ctx->data.cu.error;
+
+	/* Enqueue notification towards LL */
+	ll_rx_enqueue(ntf);
+}
+
+static void rp_cu_conn_param_req_ntf(struct ull_cp_conn *conn, struct proc_ctx *ctx)
+{
+	struct node_rx_pdu *ntf;
+	struct pdu_data *pdu;
+
+	/* Allocate ntf node */
+	ntf = ntf_alloc();
+	LL_ASSERT(ntf);
+
+	ntf->hdr.type = NODE_RX_TYPE_DC_PDU;
+	pdu = (struct pdu_data *) ntf->pdu;
+
+	pdu_encode_conn_param_req(ctx, pdu);
+
+	/* Enqueue notification towards LL */
+	ll_rx_enqueue(ntf);
+}
+
+static void rp_cu_complete(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt, void *param)
+{
+	if (!ntf_alloc_is_available()) {
+		ctx->state = RP_CU_STATE_WAIT_NTF;
+	} else {
+		rp_cu_ntf(conn, ctx);
+		rr_complete(conn);
+		ctx->state = RP_CU_STATE_IDLE;
+	}
+}
+
+static void rp_cu_send_conn_update_ind(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt, void *param)
+{
+	if (!tx_alloc_is_available()) {
+		ctx->state = RP_CU_STATE_WAIT_TX_CONN_UPDATE_IND;
+	} else {
+		/* TODO(thoh): Hardcoded instant delta +6 */
+		ctx->data.cu.instant = rp_event_counter(conn) + 6;
+		rp_cu_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_CONN_UPDATE_IND);
+		ctx->rx_opcode = 0xFF; /* TODO(thoh): Hmm */
+		ctx->state = RP_CU_STATE_WAIT_INSTANT;
+	}
+}
+
+static void rp_cu_send_reject_ind(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt, void *param)
+{
+	if (!tx_alloc_is_available()) {
+		ctx->state = RP_CU_STATE_WAIT_TX_REJECT_IND;
+	} else {
+		rp_cu_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_REJECT_IND);
+		rr_complete(conn);
+		ctx->state = RP_CU_STATE_IDLE;
+	}
+}
+
+static void rp_cu_send_conn_param_rsp(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt, void *param)
+{
+	if (!tx_alloc_is_available()) {
+		ctx->state = RP_CU_STATE_WAIT_TX_CONN_PARAM_RSP;
+	} else {
+		rp_cu_tx(conn, ctx, PDU_DATA_LLCTRL_TYPE_CONN_PARAM_RSP);
+		ctx->rx_opcode = PDU_DATA_LLCTRL_TYPE_CONN_UPDATE_IND;
+		ctx->state = RP_CU_STATE_WAIT_RX_CONN_UPDATE_IND;
+	}
+}
+
+static void rp_cu_send_conn_param_req_ntf(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt, void *param)
+{
+	if (!ntf_alloc_is_available()) {
+		ctx->state = RP_CU_STATE_WAIT_NTF_CONN_PARAM_REQ;
+	} else {
+		rp_cu_conn_param_req_ntf(conn, ctx);
+		ctx->state = RP_CU_STATE_WAIT_CONN_PARAM_REQ_REPLY;
+	}
+}
+
+static void rp_cu_st_idle(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt, void *param)
+{
+	/* TODO */
+	switch (evt) {
+	case RP_CU_EVT_RUN:
+		ctx->state = RP_CU_STATE_WAIT_RX_CONN_PARAM_REQ;
+		break;
+	default:
+		/* Ignore other evts */
+		break;
+	}
+}
+
+static void rp_cu_st_wait_rx_conn_param_req(struct ull_cp_conn *conn, struct proc_ctx *ctx, u8_t evt, void *param)
```

Comment:
At some point we'll have to check the parameters in the procedure, and send a BT_HCI_ERR_INVALID_LL_PARAM if they are invalid. Would this be the place for that? Might be good to add a comment as a placeholder for that